File: legacy/matching/scheduler.go
package matching

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"pharmabroker/domain/entity"
	"pharmabroker/pkg/config"

	"github.com/robfig/cron/v3"
)

// LearningScheduler manages scheduled weight learning jobs
type LearningScheduler struct {
	learner *WeightLearner
	config  config.AdaptiveLearningConfig
	cron    *cron.Cron
	logger  *slog.Logger

	// State tracking
	mu            sync.RWMutex
	lastRun       time.Time
	lastStatus    JobStatus
	lastError     error
	lastMetrics   *entity.PerformanceMetrics
	pendingApply  *Weights // Calculated but not applied weights
	pendingReason string   // Why pending (e.g., "manual review required")
}

// JobStatus represents the status of the last learning job
type JobStatus string

const (
	JobStatusPending     JobStatus = "pending"
	JobStatusRunning     JobStatus = "running"
	JobStatusSuccess     JobStatus = "success"
	JobStatusFailed      JobStatus = "failed"
	JobStatusSkipped     JobStatus = "skipped"
	JobStatusRecommended JobStatus = "recommended" // Weights calculated but not applied
)

// NewLearningScheduler creates a new learning scheduler
func NewLearningScheduler(
	learner *WeightLearner,
	cfg config.AdaptiveLearningConfig,
	logger *slog.Logger,
) *LearningScheduler {
	if logger == nil {
		logger = slog.Default()
	}

	return &LearningScheduler{
		learner:    learner,
		config:     cfg,
		logger:     logger,
		lastStatus: JobStatusPending,
	}
}

// Start begins the scheduled learning jobs
func (ls *LearningScheduler) Start() error {
	if !ls.config.Enabled {
		ls.logger.Info("Adaptive learning is disabled")
		return nil
	}

	ls.cron = cron.New()

	_, err := ls.cron.AddFunc(ls.config.Schedule, ls.runJob)
	if err != nil {
		return fmt.Errorf("invalid cron schedule '%s': %w", ls.config.Schedule, err)
	}

	ls.cron.Start()
	ls.logger.Info("Learning scheduler started",
		"schedule", ls.config.Schedule,
		"auto_apply", ls.config.AutoApply.Enabled,
	)

	return nil
}

// Stop gracefully stops the scheduler
func (ls *LearningScheduler) Stop() {
	if ls.cron != nil {
		ctx := ls.cron.Stop()
		<-ctx.Done()
		ls.logger.Info("Learning scheduler stopped")
	}
}

// RunNow triggers an immediate learning job (for manual/testing)
func (ls *LearningScheduler) RunNow() error {
	ls.runJob()

	ls.mu.RLock()
	defer ls.mu.RUnlock()

	if ls.lastError != nil {
		return ls.lastError
	}
	return nil
}

// runJob is the main job execution function
func (ls *LearningScheduler) runJob() {
	ls.mu.Lock()
	ls.lastRun = time.Now()
	ls.lastStatus = JobStatusRunning
	ls.lastError = nil
	ls.mu.Unlock()

	ctx := context.Background()
	logger := ls.logger.With("job", "learning", "started_at", ls.lastRun)

	logger.Info("Learning job started")

	// Calculate date range from config
	endDate := time.Now()
	startDate := endDate.Add(-time.Duration(ls.config.Algorithm.AnalysisWindowDays) * 24 * time.Hour)

	// Calculate optimal weights
	newWeights, newMetrics, err := ls.learner.CalculateOptimalWeights(ctx, startDate, endDate)
	if err != nil {
		ls.handleError(err, logger)
		return
	}

	logger.Info("Weights calculated",
		"sample_size", newMetrics.SampleSize,
		"confirmation_rate", fmt.Sprintf("%.2f%%", newMetrics.ConfirmationRate*100),
	)

	// Get current weights for comparison
	currentWeights := ls.learner.scorer.GetWeights()
	currentMetrics := ls.getCurrentMetrics(ctx)

	// Decide whether to apply
	shouldApply := ls.shouldApply(currentMetrics, *newMetrics)

	if shouldApply && ls.config.AutoApply.Enabled {
		// Auto-apply the weights
		err = ls.learner.ApplyWeights(
			ctx,
			*newWeights,
			entity.WeightSourceAutoLearned,
			newMetrics,
			ls.buildNote(currentWeights, *newWeights, *newMetrics),
		)
		if err != nil {
			ls.handleError(err, logger)
			return
		}

		ls.mu.Lock()
		ls.lastStatus = JobStatusSuccess
		ls.lastMetrics = newMetrics
		ls.pendingApply = nil
		ls.pendingReason = ""
		ls.mu.Unlock()

		logger.Info("Weights applied successfully",
			"old_weights", currentWeights,
			"new_weights", *newWeights,
		)

		if ls.config.Notifications.OnSuccess {
			ls.notifySuccess(*newWeights, *newMetrics)
		}
	} else {
		// Store for manual review
		reason := ls.getSkipReason(shouldApply)

		ls.mu.Lock()
		ls.lastStatus = JobStatusRecommended
		ls.lastMetrics = newMetrics
		ls.pendingApply = newWeights
		ls.pendingReason = reason
		ls.mu.Unlock()

		logger.Info("Weights calculated but not applied",
			"reason", reason,
			"recommended_weights", *newWeights,
		)

		if ls.config.Notifications.OnRecommendation {
			ls.notifyRecommendation(*newWeights, *newMetrics, reason)
		}
	}
}

// shouldApply determines if weights should be auto-applied
func (ls *LearningScheduler) shouldApply(current, new entity.PerformanceMetrics) bool {
	if !ls.config.AutoApply.RequireImprovement {
		return true
	}

	// Calculate separation (confirmed - rejected avg scores)
	currentSep := current.AvgScoreConfirmed - current.AvgScoreRejected
	newSep := new.AvgScoreConfirmed - new.AvgScoreRejected

	// Check separation gain
	sepGain := newSep - currentSep
	if sepGain < ls.config.AutoApply.MinSeparationGain {
		return false
	}

	// Check confirmation rate drop
	rateDrop := current.ConfirmationRate - new.ConfirmationRate
	return rateDrop <= ls.config.AutoApply.MaxConfirmationRateDrop
}

// getSkipReason explains why weights weren't applied
func (ls *LearningScheduler) getSkipReason(shouldApply bool) string {
	if !ls.config.AutoApply.Enabled {
		return "auto-apply disabled, manual review required"
	}
	if !shouldApply {
		return "performance improvement threshold not met"
	}
	return "unknown"
}

// handleError processes errors and updates state
func (ls *LearningScheduler) handleError(err error, logger *slog.Logger) {
	ls.mu.Lock()
	ls.lastStatus = JobStatusFailed
	ls.lastError = err
	ls.mu.Unlock()

	logger.Error("Learning job failed", "error", err)

	if ls.config.Notifications.OnFailure {
		ls.notifyFailure(err)
	}
}

// getCurrentMetrics fetches or estimates current performance metrics
func (ls *LearningScheduler) getCurrentMetrics(ctx context.Context) entity.PerformanceMetrics {
	// Try to get from last run or estimate from current data
	ls.mu.RLock()
	if ls.lastMetrics != nil {
		metrics := *ls.lastMetrics
		ls.mu.RUnlock()
		return metrics
	}
	ls.mu.RUnlock()

	// Return neutral metrics if no history
	return entity.PerformanceMetrics{
		ConfirmationRate:  0.5,
		AvgScoreConfirmed: 0.7,
		AvgScoreRejected:  0.3,
	}
}

// buildNote creates a descriptive note for the weight change
func (ls *LearningScheduler) buildNote(old, new Weights, metrics entity.PerformanceMetrics) string {
	return fmt.Sprintf(
		"Auto-learned from %d samples. Separation: %.3f. "+
			"Weights changed: med %.2f→%.2f, dosage %.2f→%.2f, qty %.2f→%.2f, price %.2f→%.2f, recency %.2f→%.2f",
		metrics.SampleSize,
		metrics.AvgScoreConfirmed-metrics.AvgScoreRejected,
		old.Medication, new.Medication,
		old.Dosage, new.Dosage,
		old.Quantity, new.Quantity,
		old.Price, new.Price,
		old.Recency, new.Recency,
	)
}

// Notification stubs (can be expanded with actual notification implementation)
func (ls *LearningScheduler) notifySuccess(weights Weights, metrics entity.PerformanceMetrics) {
	ls.logger.Info("NOTIFICATION: Weights applied",
		"new_weights", weights,
		"metrics", metrics,
	)
}

func (ls *LearningScheduler) notifyFailure(err error) {
	ls.logger.Error("NOTIFICATION: Learning failed",
		"error", err,
	)
}

func (ls *LearningScheduler) notifyRecommendation(weights Weights, metrics entity.PerformanceMetrics, reason string) {
	ls.logger.Info("NOTIFICATION: New weights recommended",
		"recommended_weights", weights,
		"metrics", metrics,
		"reason", reason,
	)
}

// Status returns the current scheduler status
func (ls *LearningScheduler) Status() SchedulerStatus {
	ls.mu.RLock()
	defer ls.mu.RUnlock()

	return SchedulerStatus{
		Enabled:       ls.config.Enabled,
		Schedule:      ls.config.Schedule,
		LastRun:       ls.lastRun,
		LastStatus:    ls.lastStatus,
		LastError:     ls.lastError,
		LastMetrics:   ls.lastMetrics,
		PendingApply:  ls.pendingApply,
		PendingReason: ls.pendingReason,
	}
}

// SchedulerStatus provides current scheduler state
type SchedulerStatus struct {
	Enabled       bool
	Schedule      string
	LastRun       time.Time
	LastStatus    JobStatus
	LastError     error
	LastMetrics   *entity.PerformanceMetrics
	PendingApply  *Weights
	PendingReason string
}

// ApplyPending manually applies pending weights
func (ls *LearningScheduler) ApplyPending(ctx context.Context) error {
	ls.mu.Lock()
	pending := ls.pendingApply
	ls.mu.Unlock()

	if pending == nil {
		return fmt.Errorf("no pending weights to apply")
	}

	metrics := ls.getCurrentMetrics(ctx)

	err := ls.learner.ApplyWeights(
		ctx,
		*pending,
		entity.WeightSourceManual, // Since manually triggered
		&metrics,
		"Manually approved from scheduler recommendation",
	)
	if err != nil {
		return err
	}

	ls.mu.Lock()
	ls.pendingApply = nil
	ls.pendingReason = ""
	ls.lastStatus = JobStatusSuccess
	ls.mu.Unlock()

	ls.logger.Info("Pending weights applied manually", "weights", pending)

	return nil
}

// RejectPending clears pending weights without applying
func (ls *LearningScheduler) RejectPending() {
	ls.mu.Lock()
	defer ls.mu.Unlock()

	ls.pendingApply = nil
	ls.pendingReason = "rejected by user"
	ls.lastStatus = JobStatusSkipped

	ls.logger.Info("Pending weights rejected")
}

// UpdateConfig updates the scheduler configuration
func (ls *LearningScheduler) UpdateConfig(cfg config.AdaptiveLearningConfig) error {
	ls.mu.Lock()
	defer ls.mu.Unlock()

	ls.config = cfg

	// Also update learner config
	ls.learner.SetConfig(LearningConfig{
		LearningRate:   cfg.Algorithm.LearningRate,
		MinWeight:      cfg.Algorithm.MinWeight,
		MaxWeight:      cfg.Algorithm.MaxWeight,
		MinChange:      cfg.Algorithm.MinChange,
		MinSamples:     cfg.Algorithm.MinSamples,
		AnalysisWindow: cfg.Algorithm.AnalysisWindowDays,
	})

	ls.logger.Info("Scheduler config updated", "config", cfg)

	return nil
}

// Rollback reverts to the previous weight configuration
func (ls *LearningScheduler) Rollback(ctx context.Context) error {
	if ls.learner == nil {
		return fmt.Errorf("learner not configured")
	}

	err := ls.learner.Rollback(ctx)
	if err != nil {
		ls.logger.Error("Rollback failed", "error", err)
		return err
	}

	ls.mu.Lock()
	ls.lastStatus = JobStatusSuccess
	ls.pendingApply = nil
	ls.pendingReason = "rolled back to previous weights"
	ls.mu.Unlock()

	ls.logger.Info("Weights rolled back to previous configuration")

	return nil
}

// ApplyWeightsManual applies weights directly with manual source
func (ls *LearningScheduler) ApplyWeightsManual(ctx context.Context, weights Weights, notes string) error {
	if ls.learner == nil {
		return fmt.Errorf("learner not configured")
	}

	// Get current metrics for context
	metrics := ls.getCurrentMetrics(ctx)

	err := ls.learner.ApplyWeights(
		ctx,
		weights,
		entity.WeightSourceManual,
		&metrics,
		notes,
	)
	if err != nil {
		return err
	}

	ls.mu.Lock()
	ls.lastStatus = JobStatusSuccess
	ls.pendingApply = nil
	ls.pendingReason = ""
	ls.mu.Unlock()

	ls.logger.Info("Manual weights applied", "weights", weights, "notes", notes)

	return nil
}

// GetLearner returns the underlying weight learner for direct operations
func (ls *LearningScheduler) GetLearner() *WeightLearner {
	return ls.learner
}

File: core/src/matching/scheduler.rs
//! Learning scheduler module
//!
//! Ported from legacy/matching/scheduler.go

use chrono::{DateTime, Duration, Utc};
use std::sync::RwLock;

use super::{LearningConfig, PerformanceMetrics, WeightLearner, Weights};

/// Job execution status
/// Ported from Go: JobStatus (scheduler.go:34-43)
#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub enum JobStatus {
    /// Initial state
    #[default]
    Pending,
    /// Currently executing
    Running,
    /// Completed successfully
    Success,
    /// Execution failed
    Failed,
    /// Skipped (e.g., insufficient data)
    Skipped,
    /// Weights calculated but not applied (manual review)
    Recommended,
}

/// Auto-apply configuration
/// Ported from Go: config.AutoApplyConfig
#[derive(Debug, Clone)]
pub struct AutoApplyConfig {
    /// Enable auto-apply
    pub enabled: bool,
    /// Require improvement to apply
    pub require_improvement: bool,
    /// Minimum separation gain required
    pub min_separation_gain: f64,
    /// Maximum confirmation rate drop allowed
    pub max_confirmation_rate_drop: f64,
}

impl Default for AutoApplyConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            require_improvement: true,
            min_separation_gain: 0.01,
            max_confirmation_rate_drop: 0.05,
        }
    }
}

/// Notification configuration
#[derive(Debug, Clone, Default)]
pub struct NotificationConfig {
    pub on_success: bool,
    pub on_failure: bool,
    pub on_recommendation: bool,
}

/// Scheduler configuration
/// Ported from Go: config.AdaptiveLearningConfig
#[derive(Debug, Clone)]
pub struct SchedulerConfig {
    /// Enable scheduler
    pub enabled: bool,
    /// Cron schedule (e.g., "0 3 * * *")
    pub schedule: String,
    /// Learning algorithm config
    pub algorithm: LearningConfig,
    /// Auto-apply settings
    pub auto_apply: AutoApplyConfig,
    /// Notification settings
    pub notifications: NotificationConfig,
}

impl Default for SchedulerConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            schedule: "0 3 * * *".to_string(), // Daily at 3 AM
            algorithm: LearningConfig::default(),
            auto_apply: AutoApplyConfig::default(),
            notifications: NotificationConfig::default(),
        }
    }
}

/// Current scheduler state
/// Ported from Go: SchedulerStatus (scheduler.go:315-324)
#[derive(Debug, Clone)]
pub struct SchedulerStatus {
    pub enabled: bool,
    pub schedule: String,
    pub last_run: Option<DateTime<Utc>>,
    pub last_status: JobStatus,
    pub last_error: Option<String>,
    pub last_metrics: Option<PerformanceMetrics>,
    pub pending_apply: Option<Weights>,
    pub pending_reason: Option<String>,
}

/// Learning scheduler for automated weight optimization
/// Ported from Go: LearningScheduler (scheduler.go:17-31)
pub struct LearningScheduler {
    config: RwLock<SchedulerConfig>,
    learner: Option<WeightLearner>,
    // State tracking
    last_run: RwLock<Option<DateTime<Utc>>>,
    last_status: RwLock<JobStatus>,
    last_error: RwLock<Option<String>>,
    last_metrics: RwLock<Option<PerformanceMetrics>>,
    pending_apply: RwLock<Option<Weights>>,
    pending_reason: RwLock<Option<String>>,
}

impl Default for LearningScheduler {
    fn default() -> Self {
        Self::new(SchedulerConfig::default())
    }
}

impl LearningScheduler {
    /// Create a new learning scheduler
    pub fn new(config: SchedulerConfig) -> Self {
        Self {
            config: RwLock::new(config),
            learner: Some(WeightLearner::new()),
            last_run: RwLock::new(None),
            last_status: RwLock::new(JobStatus::Pending),
            last_error: RwLock::new(None),
            last_metrics: RwLock::new(None),
            pending_apply: RwLock::new(None),
            pending_reason: RwLock::new(None),
        }
    }

    /// Create scheduler with custom learner
    pub fn with_learner(config: SchedulerConfig, learner: WeightLearner) -> Self {
        Self {
            config: RwLock::new(config),
            learner: Some(learner),
            last_run: RwLock::new(None),
            last_status: RwLock::new(JobStatus::Pending),
            last_error: RwLock::new(None),
            last_metrics: RwLock::new(None),
            pending_apply: RwLock::new(None),
            pending_reason: RwLock::new(None),
        }
    }

    /// Check if scheduler is enabled
    pub fn is_enabled(&self) -> bool {
        self.config.read().unwrap().enabled
    }

    /// Get current status
    /// Ported from Go: LearningScheduler.Status (scheduler.go:297-312)
    pub fn status(&self) -> SchedulerStatus {
        let config = self.config.read().unwrap();
        SchedulerStatus {
            enabled: config.enabled,
            schedule: config.schedule.clone(),
            last_run: *self.last_run.read().unwrap(),
            last_status: *self.last_status.read().unwrap(),
            last_error: self.last_error.read().unwrap().clone(),
            last_metrics: self.last_metrics.read().unwrap().clone(),
            pending_apply: self.pending_apply.read().unwrap().clone(),
            pending_reason: self.pending_reason.read().unwrap().clone(),
        }
    }

    /// Determine if new weights should be auto-applied
    /// Ported from Go: LearningScheduler.shouldApply (scheduler.go:195-214)
    pub fn should_apply(&self, current: &PerformanceMetrics, new: &PerformanceMetrics) -> bool {
        let config = self.config.read().unwrap();

        if !config.auto_apply.require_improvement {
            return true;
        }

        // Calculate separation (confirmed - rejected avg scores)
        let current_sep = current.avg_score_confirmed - current.avg_score_rejected;
        let new_sep = new.avg_score_confirmed - new.avg_score_rejected;

        // Check separation gain
        let sep_gain = new_sep - current_sep;
        if sep_gain < config.auto_apply.min_separation_gain {
            return false;
        }

        // Check confirmation rate drop
        let rate_drop = current.confirmation_rate - new.confirmation_rate;
        rate_drop <= config.auto_apply.max_confirmation_rate_drop
    }

    /// Get skip reason
    /// Ported from Go: LearningScheduler.getSkipReason (scheduler.go:216-225)
    pub fn get_skip_reason(&self, should_apply: bool) -> String {
        let config = self.config.read().unwrap();
        if !config.auto_apply.enabled {
            "auto-apply disabled, manual review required".to_string()
        } else if !should_apply {
            "performance improvement threshold not met".to_string()
        } else {
            "unknown".to_string()
        }
    }

    /// Build note for weight change
    /// Ported from Go: LearningScheduler.buildNote (scheduler.go:260-273)
    pub fn build_note(&self, old: &Weights, new: &Weights, metrics: &PerformanceMetrics) -> String {
        format!(
            "Auto-learned from {} samples. Separation: {:.3}. \
             Weights changed: med {:.2}→{:.2}, dosage {:.2}→{:.2}, \
             qty {:.2}→{:.2}, price {:.2}→{:.2}, recency {:.2}→{:.2}",
            metrics.sample_size,
            metrics.avg_score_confirmed - metrics.avg_score_rejected,
            old.medication,
            new.medication,
            old.dosage,
            new.dosage,
            old.quantity,
            new.quantity,
            old.price,
            new.price,
            old.recency,
            new.recency,
        )
    }

    /// Get current metrics (for comparison)
    /// Ported from Go: LearningScheduler.getCurrentMetrics (scheduler.go:241-258)
    pub fn get_current_metrics(&self) -> PerformanceMetrics {
        if let Some(metrics) = self.last_metrics.read().unwrap().as_ref() {
            return metrics.clone();
        }

        // Default neutral metrics
        PerformanceMetrics {
            confirmation_rate: 0.5,
            avg_score_confirmed: 0.7,
            avg_score_rejected: 0.3,
            ..Default::default()
        }
    }

    /// Record job start
    fn record_job_start(&self) {
        *self.last_run.write().unwrap() = Some(Utc::now());
        *self.last_status.write().unwrap() = JobStatus::Running;
        *self.last_error.write().unwrap() = None;
    }

    /// Record job success
    fn record_success(&self, metrics: PerformanceMetrics) {
        *self.last_status.write().unwrap() = JobStatus::Success;
        *self.last_metrics.write().unwrap() = Some(metrics);
        *self.pending_apply.write().unwrap() = None;
        *self.pending_reason.write().unwrap() = None;
    }

    /// Record job as recommended (needs manual review)
    fn record_recommended(&self, metrics: PerformanceMetrics, weights: Weights, reason: String) {
        *self.last_status.write().unwrap() = JobStatus::Recommended;
        *self.last_metrics.write().unwrap() = Some(metrics);
        *self.pending_apply.write().unwrap() = Some(weights);
        *self.pending_reason.write().unwrap() = Some(reason);
    }

    /// Record job failure
    fn record_failure(&self, error: String) {
        *self.last_status.write().unwrap() = JobStatus::Failed;
        *self.last_error.write().unwrap() = Some(error);
    }

    /// Get analysis date range
    pub fn get_analysis_range(&self) -> (DateTime<Utc>, DateTime<Utc>) {
        let config = self.config.read().unwrap();
        let end_date = Utc::now();
        let start_date = end_date - Duration::days(config.algorithm.analysis_window);
        (start_date, end_date)
    }

    /// Reject pending weights without applying
    /// Ported from Go: LearningScheduler.RejectPending (scheduler.go:360-370)
    pub fn reject_pending(&self) {
        *self.pending_apply.write().unwrap() = None;
        *self.pending_reason.write().unwrap() = Some("rejected by user".to_string());
        *self.last_status.write().unwrap() = JobStatus::Skipped;

        tracing::info!("Pending weights rejected");
    }

    /// Update scheduler configuration
    /// Ported from Go: LearningScheduler.UpdateConfig (scheduler.go:372-392)
    pub fn update_config(&self, config: SchedulerConfig) {
        // Update learner config if present
        if let Some(ref learner) = self.learner {
            learner.set_config(config.algorithm.clone());
        }

        *self.config.write().unwrap() = config;

        tracing::info!("Scheduler config updated");
    }

    /// Get learner reference
    pub fn learner(&self) -> Option<&WeightLearner> {
        self.learner.as_ref()
    }

    /// Check if there are pending weights
    pub fn has_pending(&self) -> bool {
        self.pending_apply.read().unwrap().is_some()
    }

    /// Get pending weights if any
    pub fn pending_weights(&self) -> Option<Weights> {
        self.pending_apply.read().unwrap().clone()
    }
}

// ============================================================================
// Tests - Ported from scheduler_test.go
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new_learning_scheduler() {
        let config = SchedulerConfig {
            enabled: false,
            schedule: "0 3 * * *".to_string(),
            ..Default::default()
        };

        let scheduler = LearningScheduler::new(config);

        assert!(!scheduler.is_enabled());
        assert_eq!(*scheduler.last_status.read().unwrap(), JobStatus::Pending);
    }

    #[test]
    fn test_scheduler_status() {
        let config = SchedulerConfig {
            enabled: true,
            schedule: "0 3 * * *".to_string(),
            ..Default::default()
        };

        let scheduler = LearningScheduler::new(config);
        let status = scheduler.status();

        assert!(status.enabled);
        assert_eq!(status.schedule, "0 3 * * *");
        assert_eq!(status.last_status, JobStatus::Pending);
    }

    #[test]
    fn test_should_apply_require_improvement_disabled() {
        let config = SchedulerConfig {
            auto_apply: AutoApplyConfig {
                enabled: true,
                require_improvement: false,
                ..Default::default()
            },
            ..Default::default()
        };

        let scheduler = LearningScheduler::new(config);

        let old = PerformanceMetrics::default();
        let new = PerformanceMetrics::default();

        assert!(scheduler.should_apply(&old, &new));
    }

    #[test]
    fn test_should_apply_separation_improved() {
        let config = SchedulerConfig {
            auto_apply: AutoApplyConfig {
                enabled: true,
                require_improvement: true,
                min_separation_gain: 0.01,
                max_confirmation_rate_drop: 0.05,
            },
            ..Default::default()
        };

        let scheduler = LearningScheduler::new(config);

        let old = PerformanceMetrics {
            avg_score_confirmed: 0.85,
            avg_score_rejected: 0.65,
            confirmation_rate: 0.75,
            ..Default::default()
        };

        let new = PerformanceMetrics {
            avg_score_confirmed: 0.88,
            avg_score_rejected: 0.62,
            confirmation_rate: 0.76,
            ..Default::default()
        };

        assert!(scheduler.should_apply(&old, &new));
    }

    #[test]
    fn test_should_apply_insufficient_gain() {
        let config = SchedulerConfig {
            auto_apply: AutoApplyConfig {
                enabled: true,
                require_improvement: true,
                min_separation_gain: 0.10, // Need 10% gain
                max_confirmation_rate_drop: 0.05,
            },
            ..Default::default()
        };

        let scheduler = LearningScheduler::new(config);

        let old = PerformanceMetrics {
            avg_score_confirmed: 0.85,
            avg_score_rejected: 0.65,
            confirmation_rate: 0.75,
            ..Default::default()
        };

        let new = PerformanceMetrics {
            avg_score_confirmed: 0.86,
            avg_score_rejected: 0.64, // Only 0.02 gain
            confirmation_rate: 0.75,
            ..Default::default()
        };

        assert!(!scheduler.should_apply(&old, &new));
    }

    #[test]
    fn test_should_apply_rate_dropped() {
        let config = SchedulerConfig {
            auto_apply: AutoApplyConfig {
                enabled: true,
                require_improvement: true,
                min_separation_gain: 0.01,
                max_confirmation_rate_drop: 0.05,
            },
            ..Default::default()
        };

        let scheduler = LearningScheduler::new(config);

        let old = PerformanceMetrics {
            avg_score_confirmed: 0.80,
            avg_score_rejected: 0.60,
            confirmation_rate: 0.80,
            ..Default::default()
        };

        let new = PerformanceMetrics {
            avg_score_confirmed: 0.90,
            avg_score_rejected: 0.55,
            confirmation_rate: 0.70, // Dropped by 0.10
            ..Default::default()
        };

        assert!(!scheduler.should_apply(&old, &new));
    }

    #[test]
    fn test_reject_pending() {
        let scheduler = LearningScheduler::default();

        // Set pending weights
        *scheduler.pending_apply.write().unwrap() = Some(Weights {
            medication: 0.5,
            ..Default::default()
        });
        *scheduler.pending_reason.write().unwrap() = Some("test".to_string());

        scheduler.reject_pending();

        assert!(scheduler.pending_apply.read().unwrap().is_none());
        assert_eq!(*scheduler.last_status.read().unwrap(), JobStatus::Skipped);
    }

    #[test]
    fn test_build_note() {
        let scheduler = LearningScheduler::default();

        let old = Weights {
            medication: 0.45,
            dosage: 0.10,
            quantity: 0.20,
            price: 0.15,
            recency: 0.10,
        };

        let new = Weights {
            medication: 0.48,
            dosage: 0.10,
            quantity: 0.18,
            price: 0.14,
            recency: 0.10,
        };

        let metrics = PerformanceMetrics {
            sample_size: 200,
            avg_score_confirmed: 0.88,
            avg_score_rejected: 0.65,
            ..Default::default()
        };

        let note = scheduler.build_note(&old, &new, &metrics);

        assert!(!note.is_empty());
        assert!(note.contains("200"));
    }

    #[test]
    fn test_get_skip_reason_auto_apply_disabled() {
        let config = SchedulerConfig {
            auto_apply: AutoApplyConfig {
                enabled: false,
                ..Default::default()
            },
            ..Default::default()
        };

        let scheduler = LearningScheduler::new(config);

        let reason = scheduler.get_skip_reason(true);
        assert!(reason.contains("manual review"));
    }

    #[test]
    fn test_get_skip_reason_threshold_not_met() {
        let config = SchedulerConfig {
            auto_apply: AutoApplyConfig {
                enabled: true,
                ..Default::default()
            },
            ..Default::default()
        };

        let scheduler = LearningScheduler::new(config);

        let reason = scheduler.get_skip_reason(false);
        assert!(reason.contains("threshold"));
    }

    #[test]
    fn test_update_config() {
        let scheduler = LearningScheduler::default();

        let new_config = SchedulerConfig {
            enabled: true,
            schedule: "0 6 * * *".to_string(),
            algorithm: LearningConfig {
                learning_rate: 0.05,
                min_samples: 200,
                ..Default::default()
            },
            ..Default::default()
        };

        scheduler.update_config(new_config);

        let config = scheduler.config.read().unwrap();
        assert!(config.enabled);
        assert_eq!(config.schedule, "0 6 * * *");
        assert!((config.algorithm.learning_rate - 0.05).abs() < 0.001);
    }

    #[test]
    fn test_get_current_metrics_default() {
        let scheduler = LearningScheduler::default();

        let metrics = scheduler.get_current_metrics();

        // Should return neutral defaults
        assert!((metrics.confirmation_rate - 0.5).abs() < 0.001);
        assert!((metrics.avg_score_confirmed - 0.7).abs() < 0.001);
        assert!((metrics.avg_score_rejected - 0.3).abs() < 0.001);
    }

    #[test]
    fn test_record_job_states() {
        let scheduler = LearningScheduler::default();

        // Test record start
        scheduler.record_job_start();
        assert_eq!(*scheduler.last_status.read().unwrap(), JobStatus::Running);
        assert!(scheduler.last_run.read().unwrap().is_some());

        // Test record success
        let metrics = PerformanceMetrics {
            sample_size: 100,
            ..Default::default()
        };
        scheduler.record_success(metrics.clone());
        assert_eq!(*scheduler.last_status.read().unwrap(), JobStatus::Success);

        // Test record failure
        scheduler.record_failure("test error".to_string());
        assert_eq!(*scheduler.last_status.read().unwrap(), JobStatus::Failed);
        assert!(scheduler.last_error.read().unwrap().is_some());

        // Test record recommended
        let weights = Weights::default();
        scheduler.record_recommended(metrics, weights, "needs review".to_string());
        assert_eq!(
            *scheduler.last_status.read().unwrap(),
            JobStatus::Recommended
        );
        assert!(scheduler.pending_apply.read().unwrap().is_some());
    }
}
